### PR TITLE
Add podcasts

### PR DESCRIPTION
**Pull request by joshr4**

### CODE REVIEW

File: db/migrations/20230821214555_add_podcasts_and_episodes.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      table.string("user_id").notNullable().index("idx_podcast_user_id");
+      table.foreign("user_id").references("user.id");
+      table.string("name").notNullable();
+      table.string("artwork_url");
+      table
+        .string("podcast_url")
+        .notNullable()
+        .unique()
+        .index("idx_podcast_url");
+      table.string("description");
```
Comment (blastshielddown) [inline db/migrations/20230821214555_add_podcasts_and_episodes.js:14]: We should add a length here as the `string` type defaults to 255 chars. We use 1000 for album descriptions but could be more here if we want.

https://knexjs.org/guide/schema-builder.html#string

```suggestion
      table.string("description",  1000);
```

File: db/migrations/20230821214555_add_podcasts_and_episodes.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      table.string("twitter");
+      table.string("instagram");
+      table.string("youtube");
+      table.string("website");
+      table.string("npub");
+    })
+    .createTable("episode", function (table) {
+      table.uuid("id").primary().unique();
+      table.string("title").notNullable();
+      table.string("description");
```
Comment (blastshielddown) [inline db/migrations/20230821214555_add_podcasts_and_episodes.js:28]: Same as above.

```suggestion
      table.string("description", 1000);
```

File: .env.example [hunk 3]
```diff
@@ -21,6 +21,7 @@ AWS_S3_BUCKET_NAME=<AWS_S3_BUCKET_NAME>
 AWS_CDN_DOMAIN=<AWS_CDN_DOMAIN>
 AWS_CDN_ID=<AWS_CDN_ID>
 AWS_S3_TRACK_PREFIX=track
+AWS_S3_EPISODE_PREFIX=episode
 AWS_S3_IMAGE_PREFIX=image
 AWS_S3_RAW_PREFIX=raw
```
Comment (blastshielddown) [inline .env.example:26]: I wonder if we even need these env vars anymore. They're identical across envs and I don't see them changing ever. Maybe we just fix as constants somewhere in the codebase?

```suggestion
```

File: controllers/episodes.ts [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+  //     res.send({ success: true, data: data });
+  //   })
+  //   .catch((err) => {
+  //     log.debug(`Error querying track table for New: ${err}`);
+  //     next(err);
+  //   });
+});
+
+export const delete_episode = asyncHandler(async (req, res, next) => {
+  const { uid, episodeId } = req.params;
```
Comment (blastshielddown) [inline controllers/episodes.ts:95]: Does this work? I thought the `uid` object was added directly to the request by the auth middleware. Like `req["uid"]`

File: controllers/podcasts.ts [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+import { isPodcastOwner } from "../library/userHelper";
+import { invalidateCdn } from "../library/cloudfrontClient";
+import Sentry from "@sentry/node";
+
+const imagePrefix = `${process.env.AWS_S3_IMAGE_PREFIX}`;
+const localConvertPath = `${process.env.LOCAL_CONVERT_PATH}`;
+const cdnDomain = `${process.env.AWS_CDN_DOMAIN}`;
+
+export const get_podcasts_by_account = asyncHandler(async (req, res, next) => {
+  const { uid } = req.params;
```
Comment (blastshielddown) [inline controllers/podcasts.ts:21]: Same question as comment in `delete_episode`

File: db/migrations/20230821214555_add_podcasts_and_episodes.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      table.foreign("podcast_id").references("podcast.id");
+      table.integer("order").unsigned().notNullable();
+      table.integer("play_count").unsigned().defaultTo(0);
+      table.bigInteger("msat_total").unsigned().defaultTo(0);
+      table.string("live_url").notNullable();
+      table.string("raw_url", 128);
+      table.integer("size").unsigned();
+      table.integer("duration").unsigned();
+      table.boolean("deleted").defaultTo(false);
+      table.boolean("is_draft").defaultTo(true);
```
Comment (blastshielddown) [inline db/migrations/20230821214555_add_podcasts_and_episodes.js:39]: Should add `is_processing` field

```suggestion
      table.boolean("is_draft").defaultTo(true);
      table.boolean("is_processing").defaultTo(false);
```

File: controllers/episodes.ts (hunk 4 above)
Comment (joshr4) [reply controllers/episodes.ts:95]: oh yeah good catch, `uid` is not coming in on the params

File: db/migrations/20230821214555_add_podcasts_and_episodes.js (hunk 6 above)
Comment (joshr4) [reply db/migrations/20230821214555_add_podcasts_and_episodes.js:40]: oops i have a commit i forgot to push that has this

File: .env.example (hunk 3 above)
Comment (joshr4) [reply .env.example]: removed them 2a23f1ddf411d4df8c59d2c606d012874df6f1d5

File: db/migrations/20230821214555_add_podcasts_and_episodes.js (hunk 1 above)
Comment (joshr4) [reply db/migrations/20230821214555_add_podcasts_and_episodes.js]: 608133f9705eeb676d5f7e612299165b6c31cdf9

File: db/migrations/20230821214555_add_podcasts_and_episodes.js (hunk 2 above)
Comment (joshr4) [reply db/migrations/20230821214555_add_podcasts_and_episodes.js]: 608133f9705eeb676d5f7e612299165b6c31cdf9

File: controllers/episodes.ts (hunk 4 above)
Comment (joshr4) [reply controllers/episodes.ts]: i think i had copied this from the artist controller - https://github.com/wavlake/catalog/blob/main/controllers/artists.js#L46

for whatever reason, `uid` is on both `params` and on `req`

both `req.params.uid` and `req.uid` are populated with the user's id

File: prisma/schema.prisma [hunk 7]
```diff
@@ -304,3 +352,21 @@ view TrackInfo {
 
   @@map("track_info")
 }
+
+/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by Prisma Client.
```
Comment (blastshielddown) [inline prisma/schema.prisma:356]: Is there a way to fix this? Like would adding `@@map` do the trick? Just wondering why the client can handle the track view but not this one.

**Review by blastshielddown — APPROVED** [review]

File: controllers/episodes.ts (hunk 4 above)
Comment (blastshielddown) [reply controllers/episodes.ts]: This is probably handled weirdly in the auth middleware

**Pull request merged**